KSPCommunityFixes/Performance/PartParsingPerf.cs

```diff
@@ -0,0 +1,666 @@
+                    {
+                        if (first)
+                        {
+                            first = false;
+                            partBounds = renderer.bounds;
+                        }
+                        else
+                        {
+                            partBounds.Encapsulate(renderer.bounds);
+                        }
+                    }
+
+                    renderer.GetSharedMaterials(materialBuffer);
+                    for (int j = materialBuffer.Count; j-- > 0;)
+                    {
+                        Material partMaterial = materialBuffer[j];
+                        if (partMaterial.IsNullOrDestroyed())
+                            continue;
+
+                        Material iconMaterial = new Material(GetIconShader(partMaterial.shader.name));
+                        iconMaterial.name = partMaterial.name;
+                        iconMaterial.CopyPropertiesFromMaterial(partMaterial);
+                        if (iconMaterial.HasProperty(ShaderHelpers.ColorPropId))
+                        {
+                            Color originalColor = iconMaterial.color;
+                            iconMaterial.SetColor(PropertyIDs._Color, new Color(
+                                originalColor.r < 0.5f ? 0.5f : originalColor.r,
+                                originalColor.g < 0.5f ? 0.5f : originalColor.g,
+                                originalColor.b < 0.5f ? 0.5f : originalColor.b));
+                        }
+                        else
+                        {
+                            iconMaterial.SetColor(PropertyIDs._Color, Color.white);
+                        }
+
+                        iconMaterial.SetFloat(PropertyIDs._Multiplier, partLoader.shaderMultiplier);
+                        iconMaterial.SetFloat(PropertyIDs._MinX, 0f);
+                        iconMaterial.SetFloat(PropertyIDs._MaxX, 1f);
+                        iconMaterial.SetFloat(PropertyIDs._MinY, 0f);
+                        iconMaterial.SetFloat(PropertyIDs._MaxY, 1f);
+                        materialBuffer[j] = iconMaterial;
+                    }
+
+                    if (materialBuffer.Count == 1)
+                        renderer.sharedMaterial = materialBuffer[0];
+                    else
+                        renderer.sharedMaterials = materialBuffer.ToArray();
+
+                    materialBuffer.Clear();
+                }
+
+            }
+            finally
+            {
+                componentBuffer.Clear();
+                iconHiddenComponentBuffer.Clear();
+                colliderObjectsToDestroy.Clear();
+                iconRenderers.Clear();
+                materialBuffer.Clear();
+            }
+
+            Vector3 size = partBounds.size;
+            float x = Math.Abs(size.x);
+            float y = Math.Abs(size.y);
+            float z = Math.Abs(size.z);
+            iconScale = x > y ? x : y;
+            iconScale = iconScale > z ? iconScale : z;
```

Comment:
curiosity: does Math.max get optimized differently by the compiler here than this expression?

Reply:
Surprisingly, even a function as small as `Math.Max()` is not inlined by default by the rather terrible JIT compiler used by Unity/KSP, so it actually does matter, at least in hot paths.

This method is far from a hot path, but I guess I've taken the habit of manually inlining stuff like that even when not necessary, which is poor practice as it make the code a bit less idiomatic and harder to read. I should add my own versions of such methods with the aggressive inlining attribute, that would be better.

Reply:
Makes sense. The reason it stood out  to me was because you only did it for Math.Max
e.g. `Math.Abs(size.y)` could have been `size.y >= 0.0 ? size.y : -size.y`
And so on :)

Given the JIT you have to work with, I can see the argument for doing it. Though as you said, if the compiler will happily inline your own Math utility class so long as you do```[MethodImpl(MethodImplOptions.AggressiveInlining)]```, that's the best of both worlds.

Pity we can't (or maybe we could, I'd have to think about it) peephole optimize some of this in the IL before the Unity Mono JIT gets its hands on it.

Reply:
> Pity we can't (or maybe we could? I know you're making changes to IL already, just not sure what the restrictions are and why they are / whether you're doing it in place, putting in thunks, or what) peephole optimize some of this in the IL before the Unity Mono JIT gets its hands on it.

Ah... that's what "transpiler" patches are (I was thinking one language to another, not  IL to IL).
Maybe I'll try that sometime, if you haven't already. But agreed that MethodImplOptions or manual inlining is better for code you're patching in, rather than having to transpile the patched code too.